File: python/23.06/86.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        for i in range(200):
            before = None
            node = head
            ch, b_ch = None, None
            while node:
                if node.val >= x:
                    ch = node
                    b_ch = before
                    break
                before = node
                node = node.next
            if ch:
                before = ch
                node = ch.next
            while node:
                if node.val < x:
                    comp = node
                    b_comp = before
                    break
                before = node
                node = node.next
            else:
                return head
            if b_ch:
                b_ch.next = comp
                b_comp.next = comp.next
                comp.next = ch
            else:
                b_comp.next = comp.next
                comp.next = ch
                head = comp
        return head

File: typescript/23.08/86.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function partition(head: ListNode | null, x: number): ListNode | null {
    if (!head) {
        return head
    }

    let ans1 = null, ans2 = null, tail1 = null, tail2 = null
    while(head) {
        if (head.val < x) {
            if (tail1) {
                tail1.next = new ListNode(head.val, null)
                tail1 = tail1.next
            } else {
                ans1 = new ListNode(head.val, null)
                tail1 = ans1
            }
        } else {
            if (tail2) {
                tail2.next = new ListNode(head.val, null)
                tail2 = tail2.next
            } else {
                ans2 = new ListNode(head.val, null)
                tail2 = ans2
            }
        }
        head = head.next
    }

    if (tail1) {
        tail1.next = ans2
        return ans1
    }
    
    return ans2
};
